File: InvokeCS/Invoke/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace WindowsFormsApplication1
{
    public partial class Form1 : Form
    {
        /// <summary>
        /// 定義一個自訂借錢動作的委派(想像成是定義一個介面，規定參數及回傳型態就對了)
        /// </summary>
        /// <param name="amount"></param>
        /// <returns></returns>
        delegate string CustomAction(string amount);

        /// <summary>
        /// 產生委派的參考
        /// </summary>
        CustomAction customActionc;
        string x = "10";
        public Form1()
        {
            InitializeComponent();
        }

        /// <summary>
        /// 正妹來借錢了
        /// </summary>
        private void btnGirl_Click(object sender, EventArgs e)
        {
            //只要是符合委派格式的function，就可以指定給他
            customActionc = LendToGirl;
            LendAction("30萬", customActionc);
        }

        /// <summary>
        /// 死黨來借錢了
        /// </summary>
        private void btnFriend_Click(object sender, EventArgs e)
        {
            customActionc = LendToFriend;
            LendAction("100", customActionc);
        }

        /// <summary>
        /// 魯蛇來借錢了
        /// </summary>
        private void btnLoser_Click(object sender, EventArgs e)
        {
            customActionc = LendToLoser;
            LendAction("10", customActionc);
        }

        /// <summary>
        /// 借錢動作
        /// </summary>
        /// <param name="amount"></param>
        /// <param name="customAct"></param>
        private void LendAction(string amount, CustomAction customAct)
        {
            txtResult.Text = string.Empty;

            //決定要借出的金額
            string finalAmount;

            //我們不需要知道這個customAct到底是什麼
            //反正他跑完會回傳一個我們要的東西就對了
            //在這裡回傳的就是最終借出金額
            finalAmount = customAct(amount);

            string commonRes;
            if (!string.IsNullOrEmpty(finalAmount))
            {
                commonRes = string.Format("借出{0}元", finalAmount);
            }
            else
            {
                commonRes = "掉頭就走";
            }

            txtResult.Text += commonRes;
        }

        /// <summary>
        /// 借錢給正妹的自訂動作
        /// </summary>
        /// <param name="amount"></param>
        /// <returns></returns>
        private string LendToGirl(string amount)
        {
            //自訂動作：跟正妹狂聊，最後決定借五百萬
            var res =
    @"詢問正妹：真的只要借{0}嗎?夠不夠啊?
詢問正妹：要幫妳買點數卡嗎?
詢問正妹：可以加妳的Line嗎?
詢問正妹：妳幾歲呀?
詢問正妹：妳住哪?
詢問正妹：妳有男朋友嗎?
詢問正妹：妳三圍多少?
詢問正妹：禮拜六有空嗎?
...
.....
....
哇!服務這麼好喔!
....
.....
GGInInDer
OK~{1}沒問題!
....
...去提款機領{1}元
";
            var finalAmount = "五百萬";
            txtResult.Text = string.Format(res, amount, finalAmount);

            //回傳最後決定的金額
            return finalAmount;
        }

        /// <summary>
        /// 借錢給死檔的自訂動作
        /// </summary>
        /// <param name="amount"></param>
        /// <returns></returns>
        private string LendToFriend(string amount)
        {
            //自訂動作：馬上就決定
            var res =
    @"幹...
(錢包掏出{0}元)
";
            txtResult.Text = string.Format(res, amount);
            return amount;
        }

        /// <summary>
        /// 借錢給魯蛇的自訂動作
        /// </summary>
        /// <param name="amount"></param>
        /// <returns></returns>
        private string LendToLoser(string amount)
        {
            //自訂動作：什麼都不做
            return string.Empty;
        }
    }
}

File: InvokeVB/WindowsApplication1/Form1.vb
Public Class Form1
    ''' <summary>
    ''' 定義一個自訂借錢動作的委派(想像成是定義一個介面，規定參數及回傳型態就對了)
    ''' </summary>
    ''' <param name="amount"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Delegate Function CustomAction(ByVal amount As String) As String
    ''' <summary>
    ''' 產生委派的參考
    ''' </summary>
    ''' <remarks></remarks>
    Private customAct_func As CustomAction
    ''' <summary>
    ''' 正妹借錢來
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub btnGirl_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnGirl.Click
        '直接委派 LendToGirl 這個 Function 做事,逗號後面的參數則是 LendToGirl 這個 Function 所需要的引數
        'Invoke(New CustomAction(AddressOf LendToGirl), "20萬")

        '將要做事的 Function 丟入變數後，再傳到某個 Function 去開工
        customAct_func = AddressOf LendToGirl '把某個 Function 丟進 customAct_func 變數中
        LendAction("30萬", customAct_func) '這樣就可以把 Function 當成引數傳到其他 Function 中
    End Sub
    ''' <summary>
    ''' 死黨來借錢
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub btnFriend_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnFriend.Click
        customAct_func = AddressOf LendToFriend
        LendAction("100", customAct_func)
        'Invoke(New CustomAction(AddressOf LendToFriend), "100")
    End Sub
    ''' <summary>
    ''' 魯蛇來借錢
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub btnLoser_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnLoser.Click
        customAct_func = AddressOf LendToLoser
        LendAction("10", customAct_func)
        'Invoke(New CustomAction(AddressOf LendToLoser), "10")
    End Sub

    ''' <summary>
    ''' 借錢動作
    ''' </summary>
    ''' <param name="amount"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Function LendAction(ByVal amount As String, ByVal customAct_func As CustomAction)
        txtresult.Text = String.Empty
        Dim finalAmount As String = ""
        '我們不需要知道這個customAct_func到底是什麼
        '反正他跑完會回傳一個我們要的東西就對了
        '在這裡回傳的就是最終借出金額
        finalAmount = customAct_func(amount)
        Debug.Print("finalAmount:" & finalAmount)
        Dim commonRes As String = 0
        If Not String.IsNullOrEmpty(finalAmount) Then
            commonRes = String.Format("借出{0}元", finalAmount)
        Else
            commonRes = "掉頭就走"
        End If
        txtresult.Text += commonRes
        Return Nothing
    End Function


    ''' <summary>
    ''' 借錢給正妹的自訂動作
    ''' </summary>
    ''' <param name="amount"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function LendToGirl(ByVal amount As String) As String
        Dim res As String = ("真的只要借{0}元嗎，我可以借你{1}元!!" + Environment.NewLine)
        Dim finalAmount As String = "五百萬"
        txtresult.Text = String.Format(res, amount, finalAmount)
        Return finalAmount
    End Function


    ''' <summary>
    ''' 借錢給死檔的自訂動作
    ''' </summary>
    ''' <param name="amount"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function LendToFriend(ByVal amount As String) As String
        Dim res As String = "幹...掏出{0}元" + Environment.NewLine
        txtresult.Text = String.Format(res, amount)
        Return amount
    End Function

    ''' <summary>
    ''' 借錢給魯蛇的自訂動作
    ''' </summary>
    ''' <param name="amount"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function LendToLoser(ByVal amount As String) As String
        Return String.Empty
    End Function
End Class
